animator/src/main/kotlin/richpathanimator/AnimationBuilder.kt

```diff
@@ -0,0 +1,228 @@
+package richpathanimator
+
+import android.animation.ArgbEvaluator
+import android.animation.ObjectAnimator
+import android.animation.ValueAnimator
+import android.util.Log
+import android.view.animation.Interpolator
+import richpathanimator.RichPathAnimator.*
+import com.richpath.RichPath
+import com.richpath.pathparser.PathDataNode
+import com.richpath.pathparser.PathParserCompat
+
+class AnimationBuilder(private val richPathAnimator: RichPathAnimator,
+                       private vararg val paths: RichPath) {
+
+    companion object {
+        private const val DEFAULT_DURATION = 300L
+        private const val DEFAULT_START_DELAY = 0L
+    }
+
+    val animators = arrayListOf<ValueAnimator>()
+
+    var duration = DEFAULT_DURATION
+    var startDelay = DEFAULT_START_DELAY
+    var interpolator: Interpolator? = null
+    var repeatMode: RepeatMode = RepeatMode.Restart
+    var repeatCount = 0
+
+    private fun property(propertyName: String, vararg values: Float) {
+        for (path in paths) {
+            val objectAnimator = ObjectAnimator.ofFloat(path, propertyName, *values)
+            applyAnimatorProperties(objectAnimator, path)
+        }
+    }
+
+    fun andAnimate(vararg paths: RichPath): AnimationBuilder {
+        return richPathAnimator.addAnimationBuilder(*paths)
+    }
+
+    fun thenAnimate(vararg paths: RichPath): AnimationBuilder {
+        return richPathAnimator.thenAnimate(*paths)
+    }
+
+    /**
+     * Custom animation builder.
+     *
+     * @param listener the AnimationUpdateListener
+     * @param values   A set of values that the animation will animate between over time.
+     */
+    fun custom(listener: AnimationUpdateListener?, vararg values: Float) = apply {
+        for (path in paths) {
+            val valueAnimator = ValueAnimator.ofFloat(*values).apply {
+                addUpdateListener { animation ->
+                    listener?.update(path, animation.animatedValue as Float)
+                    path.onRichPathUpdatedListener?.onPathUpdated()
+                }
+            }
+            applyAnimatorProperties(valueAnimator, path)
+        }
```

Comment:
we can move `applyAnimatorProperties(valueAnimator, path)` inside `apply{}` block then can remove `val 
 valueAnimator`

like below:

```kotlin
ValueAnimator.ofFloat(*values).apply {
    addUpdateListener { animation ->
        listener?.update(path, animation.animatedValue as Float)
        path.onRichPathUpdatedListener?.onPathUpdated()
    }
    applyAnimatorProperties(this, path)
}
```

Reply:
Yes, I agree with that. You can see in 7bb9d6a.